### ISSUE TITLE
Open file descriptors when using multiple tagged loggers with a collective configuration

### ISSUE THREAD

**Issue by MartinAnon** [tinylog-org/tinylog#342]
Hi everyone,

let's say we have 20 different java applications/services running in an application server (Tomcat, etc.).
Each application has it's own valid log configuration, stored in a collective, single tinylog configuration file, which is used for the logging configuration of all applications, e.g.

**tinylog.properties**
```
writerSERVICE1 = rolling file
writerSERVICE1.tag = SERVICE1
writerSERVICE1.format = {date:yyyy-MM-dd HH:mm:ss.SSSZ} | {class-name} | {message}
writerSERVICE1.level = info
writerSERVICE1.file = /some/logpath/service1.{date:yyyyMMdd}.{count}.log
writerSERVICE1.policies = daily, size: 10mb
writerSERVICE1.backups = 5

writerSERVICE2 = rolling file
writerSERVICE2.tag = SERVICE2
writerSERVICE2.format = {date:yyyy-MM-dd HH:mm:ss.SSSZ} | {class-name} | {message}
writerSERVICE2.level = info
writerSERVICE2.file = /some/logpath/service2.{date:yyyyMMdd}.{count}.log
writerSERVICE2.policies = daily, size: 10mb
writerSERVICE2.backups = 5
[...]
writerSERVICE20 = rolling file
writerSERVICE20.tag = SERVICE20
writerSERVICE20.format = {date:yyyy-MM-dd HH:mm:ss.SSSZ} | {class-name} | {message}
writerSERVICE20.level = info
writerSERVICE20.file = /some/logpath/service20.{date:yyyyMMdd}.{count}.log
writerSERVICE20.policies = daily, size: 10mb
writerSERVICE20.backups = 5
```

Applications are mainly using their own Tag for logging, for example SERVICE1 uses the following code to log messages to its designated logfile:
**Service1Impl.java**
```
public static final TaggedLogger SERVICE1_LOGGER = Logger.tag("SERVICE1");

//[...]
SERVICE1_LOGGER1.info("Some Info");
SERVICE1_LOGGER1.error("Some error");
```

Applications are running fine with this configuration in the application server and whenever each Service is called, each logging message is stored in its own logfile, as intended.

**Open file descriptors:**
However when inspecting the open file descriptors under a Linux System, f.e. with the command
`lsof -a -p APPLICATION_SERVER_PID | grep service1 | grep .log`
what you see is up to 20 different open file descriptors for the **same** logfile.

It looks like file descriptors are opened by each service (20) for each tag (20), regardless if the tag is used or not.
So if each service already logged an initial message with the configuration above (in **tinylog.properties**), we can see up to 400 (20x20) open file descriptors, eventhough each service really only used its own Logging Tag for logging purposes.

There might be a possibility this introduces problems (Exceptions - Too many open files), because of a default maximum open files per process limit under linux. One can inspect that limit with the linux command:
```
ulimit -a
[...]
open files                      (-n) 1024
```

Is there a way, so that the file descriptors are opened only when they are really used ?
My first idea would be to either
1. raise the maximum open file per process limit (this requires a manual operation from the system administrator)
2. to split up the configuration file (tinylog.properties), so each service has its own configuration file, only with that logging tag configuration. (The latter might solve the file descriptor problem but introduces new problems ...)

In my opinion, a single configuration file seems much easier to handle than 20 different files. Also since tinylog 2.4 #201 it's possible to create a tagged logger with multiple tags that issues each log entry to all defined tags. So with multiple configuration files, you would have to configure a tag used by multiple services again and again, in multiple different files, ugh...

Do you have any other ideas ? Is there a way, so that file descriptors are only created  f.e. the first time they are used ?

Thanks in Advance !!

**Comment by pmwmedia** [comment]
tinylog creates all configured log files at startup by design for two main reasons:

- If the file system is not accessible or the log file path misconfigured, we should fail at startup. The start of the application is the moment when errors are most likely to be noticed and not overlooked.
- Some users use monitoring tools for log files. Such tools often require the present of log files,

Nevertheless, there are multiple possible solutions I have in mind for your use case. Firstly, I have understand your use case a bit better.

> Applications are mainly using their own Tag for logging, for example SERVICE1 uses the following code to log messages to its designated logfile:

Does this mean that each application always uses only one tag, or can an application use multiple tags under certain circumstances?

When do you know which service tag will be used? Before starting the application? Before issuing the first log entry? Or even  later?

**Comment by MartinAnon** [comment]
Hi,

thanks for your reply. 

> Does this mean that each application always uses only one tag, or can an application use multiple tags under certain circumstances?

Not necessarily no. An application would mainly use its own tag and its own logfile, but it is possible that it could use multiple tags, if it has a message that is interesting in other logfiles (f.e. to monitor the run-time of different services in the same log file)

> When do you know which service tag will be used? Before starting the application? Before issuing the first log entry? Or even later?

I'd say, I know what logging tags are used before issuing the first log entry and even before the application has even started.
- Before the application has started: Because services will mainly use their own logging tag. If there are exceptions from that rule, then there have to be specific reasons why. (monitor run-time, system-access, ...). 
- I'd also say, before issuing the first log entry, because in this scenario, each service has to get its Tagged Logger(s) that it wants to use.
`Logger.tag("SERVICE1");`

**Comment by pmwmedia** [comment]
> I'd also say, before issuing the first log entry, because in this scenario, each service has to get its Tagged Logger(s) that it wants to use.

In tinylog, the configuration is loaded for the entire logging framework when any logger issues the first log entry. For example, if the `SERVICE1` logger is the first logger that issues the first log entry, the entire configuration for all loggers will be loaded. Therefore, we cannot parameterized the configuration file as I have initially hoped.

However, you could create a lazy wrapper class for the rolling file writer for creating log files only on demand.

```java
import org.tinylog.core.LogEntry;
import org.tinylog.writers.AbstractFormatPatternWriter;
import org.tinylog.writers.RollingFileWriter;
import org.tinylog.writers.Writer;

import java.util.Map;

public class LazyFileWriter extends AbstractFormatPatternWriter {

	private final Map<String, String> properties;
	private volatile Writer writer;

	public LazyFileWriter(Map<String, String> properties) {
		super(properties);
		this.properties = properties;
	}

	@Override
	public void write(LogEntry logEntry) throws Exception {
		if (writer == null) {
			synchronized (properties) {
				if (writer == null) {
					writer = new RollingFileWriter(properties);
				}
			}
		}

		writer.write(logEntry);
	}

	@Override
	public void flush() throws Exception {
		Writer writer = this.writer;
		if (writer != null) {
			writer.flush();
		}
	}

	@Override
	public void close() throws Exception {
		Writer writer = this.writer;
		if (writer != null) {
			writer.close();
		}
	}

}
```

You can register this writer as service as described in https://tinylog.org/v2/extending/#custom-writer and use `lazy file` instead of `rolling file` in your configuration.

**Comment by MartinAnon** [comment]
With the code example above, I only added one more file:
- META-INF/services/org.tinylog.writers.Writer
with the content of the packageName and the className, f.e.
`com.example.LazyFileWriter`
(as described in the documentation - https://tinylog.org/v2/extending/#custom-writer)

With those changes, it looks really good. I can only see one open file descriptor for every service and it's log file, when using the lsof command from before:
`lsof -a -p APPLICATION_SERVER_PID | grep service1 | grep .log`

Therefore I marked this question as closed.
Thanks for the help and the in-depth code examples !!
